pkg/retry/retry.go

```diff
@@ -5,9 +5,199 @@
+			retrySettings.RetryMaxCount = retryMaxCount
+		} else {
+			return nilRetrySettings, err
+		}
+	} else {
+		log.Debugf("no retry max count provided. setting default %d", baseRetrySettings.RetryMaxCount)
+	}
+
+	if retryIntervalInSeconds != 0 {
+		err := validateRetryIntervalInSeconds(retryIntervalInSeconds)
+		if err == nil {
+			log.Debugf("setting retry interval %d", retryIntervalInSeconds)
+			retrySettings.RetryIntervalInSeconds = retryIntervalInSeconds
+		} else {
+			return nilRetrySettings, err
+		}
+	} else {
+		log.Debugf("no retry interval provided. setting default %s", baseRetrySettings.RetryIntervalInSeconds)
+	}
+
+	return retrySettings, nil
+}
+
+func validateRetrySettings(retrySettings RetrySettings) error {
+	err := validateRetryStrategy(string(retrySettings.RetryStrategy))
+	if err == nil {
+		err = validateRetryMaxCount(retrySettings.RetryMaxCount)
+	}
+	if err == nil {
+		err = validateRetryIntervalInSeconds(retrySettings.RetryIntervalInSeconds)
+	}
+	return err
+}
+
+func validateRetryStrategy(retryStrategy string) error {
+	for _, allowedRetryStrategy := range AllowedRetryStrategies {
+		if retryStrategy == string(allowedRetryStrategy) {
+			return nil
+		}
+	}
+	return errors.Errorf("retry strategy %s is not valid", retryStrategy)
+}
+
+func validateRetryMaxCount(retryMaxCount int) error {
+	if retryMaxCount > MaxRetryMaxCount || retryMaxCount < MinRetryMaxCount {
+		return errors.Errorf("retry max count of %d is out of range [%d-%d]", retryMaxCount, MinRetryMaxCount, MaxRetryMaxCount)
+	}
+	return nil
+}
+
+func validateRetryIntervalInSeconds(retryIntervalInSeconds int) error {
+	_, isValidDuration := time.ParseDuration(fmt.Sprintf("%ss", strconv.Itoa(retryIntervalInSeconds)))
+	if isValidDuration != nil {
+		return errors.Errorf("retry interval value provided of %d fails to convert to a valid duration with %s", retryIntervalInSeconds, isValidDuration.Error())
+	}
+	if retryIntervalInSeconds > MaxRetryIntervalInSeconds || retryIntervalInSeconds < MinRetryIntervalInSeconds {
+		return errors.Errorf("retry interval of %d is out of range [%d-%d]", retryIntervalInSeconds, MinRetryIntervalInSeconds, MaxRetryIntervalInSeconds)
+	}
+	return nil
+}
+
+// Retry executes an Operation as per the Retry Policy specified and till it succeeds or the max number of retries specified in the retry settings provided
+func Retry(operation Operation, retrySettings RetrySettings) error {
+	err := validateRetrySettings(retrySettings)
+	if err != nil {
+		return errors.Errorf("failed to execute Operation with retry due to invalid retry settings: %s", err.Error())
+	}
+	err = operation()
+	if err != nil && retrySettings.RetryStrategy != off {
+		log.Infof("Operation failed. Will retry as per the retrySettings provided: retryStrategy=%s, retryMaxCount=%d, retryIntervalInSeconds=%d", retrySettings.RetryStrategy, retrySettings.RetryMaxCount, retrySettings.RetryIntervalInSeconds)
+		ch := make(chan error, 1)
```

Comment:
Please, use the backoff library for this. See the Azure Service Bus code for pubsub in contrib repo.

Reply:
Fixed